### ISSUE TITLE
Cluster-autoscaler: AWS EC2 Spot Fleets support

### ISSUE THREAD

**Issue by mumoshu** [kubernetes/contrib#2066]
Hi, thanks for developing cluster-autoscaler!

Is this something you'd like to add to cluster-autoscaler?

With spot fleets, we can easily mix various types of spot instances into a single logical group managed by AWS to benefit from reduced infrastructure cost and reduced operational burden including bidding, choosing which type of instance type to bid, increased availability via mixing mutliple spot instance types(assuming it will be rare that we lose to all the bids to all the different spot types at once).

A spot fleet has the notion of "target capacity" where "capacity" means number of units(a.k.a InstanceWeight) required to handle your workload e.g. 1 for r3.2xlarge, 4 for r3.8xlarge.
Basically I guess we can treat 1 target capacity = 1 instance weight as unit of autoscaling in cluster-autoscaler so that we can reuse a lot of current codebase to support spot fleets.

I'm both interested in using it and if necessary implementing it.
If you have any comments, advices, questions, etc., please let me know!

**Comment by andrewsykim** [comment]
I would be interested in helping as well, though as mentioned in https://github.com/kubernetes/contrib/issues/1921, the CA needs some sort of logic for choosing a node group which I assume would be a pre-requisite for this.

**Comment by mumoshu** [comment]
Hi @andrewsykim, I've been thinking of how we'd like cluster-autoscaler to select which node pool(asg or spot fleet) to expand for a few days.

I've now realized that the current "random" strategy plus "most-pods" and "least-waste"  proposed in https://github.com/kubernetes/contrib/pull/2118 would work in most cases, as long as we choose the `diversified` strategy and launch specifications with enough "diversity" for our spot fleets, to keep our fleets up and running.

I'd rather like cluster-autoscaler to not select an autoscaling group which is suspended, or a spot fleet which is losing to bids for all the launch specifications.
However, that's an another issue IMHO.

**Comment by andrewsykim** [comment]
@mumoshu my org and I have built a new image `wattpad/cluster-autoscaler:v1.2` with the changes in https://github.com/kubernetes/contrib/pull/2118. Trying to get as much testing in as possible, it would help this issue move forward. When you have time could you test it?

**Comment by mumoshu** [comment]
@andrewsykim Thanks for sharing the image. Sure, I'd definitely like to test it!

**Comment by motymichaely** [comment]
@mumoshu How can we move this forward?

**Comment by mumoshu** [comment]
I believe this is still relevant to every Kubernetes-on-AWS user who doesn't use a third-party solution like Spotinst.

**Comment by mumoshu** [comment]
The CA project has been moved to https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler. Could anyone open an another issue that replicates this there?

**Comment by zytek** [comment]
@mumoshu It's already opened: https://github.com/kubernetes/autoscaler/issues/838
Please close this issue as I have asked for reopen by mistake.  Sorry :)
